[SPARK-12789]Support order by index and group by index

zhichao-li opened the pull request:
1. Num in Order by is treated as constant expression at the moment. I guess it would be good to enable user to specify column by index which has been supported in Hive 0.11.0 and later.
2. The index is 1-base which means the position of the projection list for `Order by` and the position of colums for `group by`.
3. For example:
   - table test (a, b, c)
   - SELECT b, c FROM test ORDER BY 1  same as SELECT b, c FROM test ORDER BY b
   - SELECT SUM(a) FROM test GROUP BY 2  same as SELECT SUM(a) FROM test GROUP BY b
   - If we `order by 0` or `group by 4`, it would throw exception in this case since the index has been out of range.

zhichao-li commented:
cc @chenghao-intel @adrian-wang

adrian-wang commented on sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:
```diff
@@ -22,8 +22,9 @@ import java.sql.Timestamp
-import org.apache.spark.sql.catalyst.analysis.FunctionRegistry
+import org.apache.spark.sql.catalyst.analysis.{UnresolvedException, FunctionRegistry}
 import org.apache.spark.sql.catalyst.errors.DialectException
+import org.apache.spark.sql.catalyst.expressions.SortOrder
```
unused?

zhichao-li replied:
I'ts for `intercept[UnresolvedException[SortOrder]]`

adrian-wang replied:
ok

SparkQA commented:
**[Test build #49276 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49276/consoleFull)** for PR 10731 at commit [`5a2270b`](https://github.com/apache/spark/commit/5a2270be260c635772fb4d46a7084eecf7166d87).

adrian-wang commented on sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:
```diff
@@ -22,8 +22,9 @@ import java.sql.Timestamp
 import org.apache.spark.AccumulatorSuite
 import org.apache.spark.sql.catalyst.DefaultParserDialect
-import org.apache.spark.sql.catalyst.analysis.FunctionRegistry
+import org.apache.spark.sql.catalyst.analysis.{UnresolvedException, FunctionRegistry}
```
import order

adrian-wang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -441,7 +442,7 @@ class Analyzer(
 
       // When resolve `SortOrder`s in Sort based on child, don't report errors as
       // we still have chance to resolve it based on grandchild
-      case s @ Sort(ordering, global, child) if child.resolved && !s.resolved =>
```
why remove it?

zhichao-li replied:
Literal is always resolved, so remove this to ensure `Order by 1` can be matched by this case checking.

yhuai replied:
But if we have `order by a` and after resolve the ordering column, we will still hit this case, which is not what we want, right?

zhichao-li replied:
ah.. right, let me add it back.

SparkQA commented:
**[Test build #49276 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49276/consoleFull)** for PR 10731 at commit [`5a2270b`](https://github.com/apache/spark/commit/5a2270be260c635772fb4d46a7084eecf7166d87).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49276/
Test FAILed.

SparkQA commented:
**[Test build #49283 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49283/consoleFull)** for PR 10731 at commit [`d5cb4e2`](https://github.com/apache/spark/commit/d5cb4e2d90b3a64ca3a28163b3b073272eada333).

SparkQA commented:
**[Test build #49283 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49283/consoleFull)** for PR 10731 at commit [`d5cb4e2`](https://github.com/apache/spark/commit/d5cb4e2d90b3a64ca3a28163b3b073272eada333).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49283/
Test FAILed.

SparkQA commented:
**[Test build #49292 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49292/consoleFull)** for PR 10731 at commit [`b72547b`](https://github.com/apache/spark/commit/b72547b04923862579935344f6cd9af1814b77f5).

SparkQA commented:
**[Test build #49292 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49292/consoleFull)** for PR 10731 at commit [`b72547b`](https://github.com/apache/spark/commit/b72547b04923862579935344f6cd9af1814b77f5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49292/
Test FAILed.

SparkQA commented:
**[Test build #49311 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49311/consoleFull)** for PR 10731 at commit [`1cb6752`](https://github.com/apache/spark/commit/1cb675262ed3b90afdff5ffd8b1bfa4965e37df4).

SparkQA commented:
**[Test build #49311 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49311/consoleFull)** for PR 10731 at commit [`1cb6752`](https://github.com/apache/spark/commit/1cb675262ed3b90afdff5ffd8b1bfa4965e37df4).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49311/
Test FAILed.

yhuai commented:
@zhichao-li It will be good if you can take a look and see if other databases (other than hive) support this. I am not sure if it is really useful.

yhuai commented:
oh, nvm. It is pretty common in other databases.

zhichao-li commented:
retest this please.

SparkQA commented:
**[Test build #49359 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49359/consoleFull)** for PR 10731 at commit [`1cb6752`](https://github.com/apache/spark/commit/1cb675262ed3b90afdff5ffd8b1bfa4965e37df4).

SparkQA commented:
**[Test build #49359 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49359/consoleFull)** for PR 10731 at commit [`1cb6752`](https://github.com/apache/spark/commit/1cb675262ed3b90afdff5ffd8b1bfa4965e37df4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49359/
Test PASSed.

SparkQA commented:
**[Test build #49383 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49383/consoleFull)** for PR 10731 at commit [`fe99e00`](https://github.com/apache/spark/commit/fe99e0043598e6af37c691aeb00d1204a37d9158).

SparkQA commented:
**[Test build #49383 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49383/consoleFull)** for PR 10731 at commit [`fe99e00`](https://github.com/apache/spark/commit/fe99e0043598e6af37c691aeb00d1204a37d9158).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49383/
Test PASSed.

yhuai commented:
a quick question. If I do `ORDER BY a, 2, b, c`, `2` means the third column?

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -438,11 +438,25 @@ class Analyzer(
       // we still have chance to resolve it based on grandchild
-      case s @ Sort(ordering, global, child) if child.resolved && !s.resolved =>
-        val newOrdering = resolveSortOrders(ordering, child, throws = false)
+      case s @ Sort(ordering, global, child) if child.resolved =>
```
Should we keep the `&& !s.resolved` in the condition and have another case to handle the case of all literals?

zhichao-li replied:
Would update that shortly. I was thinking it would be more efficient by combining those in one past.

adrian-wang commented:
order by 2 should be the second column, I think

zhichao-li commented:
yes, It's a 1-based indexing for the projection list.

zhichao-li committed:
support order by index and group by index

refactor

style didn't catched in local

style

forgot to enbale style checking for test in local

override resolved

refactor

a version

refactor

zhichao-li commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -445,6 +445,26 @@ class Analyzer(
         val newOrdering = resolveSortOrders(ordering, child, throws = false)
         Sort(newOrdering, global, child)
 
+     // Resolve the order index to be a specific column
```
@yhuai  not sure if it's the style you prefer. mind giving suggestions?

SparkQA commented:
**[Test build #49575 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49575/consoleFull)** for PR 10731 at commit [`d9f548c`](https://github.com/apache/spark/commit/d9f548c4e0c62a08de429ce9bf96b3b7aa11b792).

SparkQA commented:
**[Test build #49575 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49575/consoleFull)** for PR 10731 at commit [`d9f548c`](https://github.com/apache/spark/commit/d9f548c4e0c62a08de429ce9bf96b3b7aa11b792).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49575/
Test FAILed.

SparkQA commented:
**[Test build #49579 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49579/consoleFull)** for PR 10731 at commit [`2746e0f`](https://github.com/apache/spark/commit/2746e0f632d7799f0d49d2d4bb8719b73c1716fd).

SparkQA commented:
**[Test build #49579 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49579/consoleFull)** for PR 10731 at commit [`2746e0f`](https://github.com/apache/spark/commit/2746e0f632d7799f0d49d2d4bb8719b73c1716fd).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49579/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49583/
Test FAILed.

SparkQA commented:
**[Test build #49584 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49584/consoleFull)** for PR 10731 at commit [`acd00be`](https://github.com/apache/spark/commit/acd00be8d92c0cea72db3e63a1580b515660d61c).

SparkQA commented:
**[Test build #49584 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49584/consoleFull)** for PR 10731 at commit [`acd00be`](https://github.com/apache/spark/commit/acd00be8d92c0cea72db3e63a1580b515660d61c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49584/
Test PASSed.

hvanhovell commented:
This is similar to: https://github.com/apache/spark/pull/10052

That PR also implements this idea for ```GROUP BY``` clauses.

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -445,6 +445,26 @@ class Analyzer(
+                  |The Select List is indexed from 1 to ${orderNodes.size}""".stripMargin)
+          }
+        }
+        val newOrdering = ordering map {
```
oh, actually, I meant if we can just check if there is any integer literal. If so, we create a new `Sort`. Otherwise, we keep the old one.

adrian-wang replied:
If there are nested attributes, integer literal could be used to reference nested fields, so we the integer literal should be on top level of order list.

zhichao-li replied:
Seems like it might be more reasonable from the semantic point of view to override the `resolved` method and move the logic to resolveSortOrders.

zhichao-li commented on sql/core/src/test/scala/org/apache/spark/sql/execution/joins/InnerJoinSuite.scala:
```diff
@@ -17,6 +17,7 @@
 
 package org.apache.spark.sql.execution.joins
 
+import org.apache.spark.sql.{DataFrame, Row, SQLConf}
```
This is for passing the style check

zhichao-li commented:
@hvanhovell didn't aware of #10052, would be happy if @dereksabryfb can pick up that.

SparkQA commented:
**[Test build #49663 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49663/consoleFull)** for PR 10731 at commit [`0daa766`](https://github.com/apache/spark/commit/0daa766d03538c806175c3389106a83b0fe977e3).

SparkQA commented:
**[Test build #49663 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49663/consoleFull)** for PR 10731 at commit [`0daa766`](https://github.com/apache/spark/commit/0daa766d03538c806175c3389106a83b0fe977e3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49663/
Test FAILed.

SparkQA commented:
**[Test build #49844 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49844/consoleFull)** for PR 10731 at commit [`e61429f`](https://github.com/apache/spark/commit/e61429fec35c0f0983ff5e1bfeea11a1cef42690).

SparkQA commented:
**[Test build #49844 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49844/consoleFull)** for PR 10731 at commit [`e61429f`](https://github.com/apache/spark/commit/e61429fec35c0f0983ff5e1bfeea11a1cef42690).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49844/
Test FAILed.

zhichao-li commented:
retest this please.

SparkQA commented:
**[Test build #49864 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49864/consoleFull)** for PR 10731 at commit [`e61429f`](https://github.com/apache/spark/commit/e61429fec35c0f0983ff5e1bfeea11a1cef42690).

SparkQA commented:
**[Test build #49864 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49864/consoleFull)** for PR 10731 at commit [`e61429f`](https://github.com/apache/spark/commit/e61429fec35c0f0983ff5e1bfeea11a1cef42690).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49864/
Test PASSed.

zhichao-li committed:
refactor

zhichao-li committed:
redundant

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -446,6 +503,10 @@ class Analyzer(
 
+      case s @ Sort(ordering, global, child) if child.resolved && !s.resolved =>
+        val newOrdering = resolveSortOrders(ordering, child, throws = false)
+        Sort(newOrdering, global, child)
```
Is this rule the same as the above one?

zhichao-li replied:
oh. my bad , had been aware of this, but forgot to push out the code.

zhichao-li committed:
typo

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49972/
Test FAILed.

SparkQA commented:
**[Test build #49973 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49973/consoleFull)** for PR 10731 at commit [`1a36a2a`](https://github.com/apache/spark/commit/1a36a2a83c9fa977aa344eeb08fa544f82d80fdc).

SparkQA commented:
**[Test build #49973 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49973/consoleFull)** for PR 10731 at commit [`1a36a2a`](https://github.com/apache/spark/commit/1a36a2a83c9fa977aa344eeb08fa544f82d80fdc).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49973/
Test PASSed.

tejasapatil commented:
I would like to have this patch in trunk. If @zhichao-li is not planning to do more changes, can one of the admins review ?

zhichao-li commented:
@yhuai any more comments?

zhichao-li commented:
cc @liancheng

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -322,6 +323,62 @@ class Analyzer(
   /**
+    * Replace the index with the related attribute for order by and group by.
+    */
+  object ResolveIndexReferences extends Rule[LogicalPlan] {
```
Do we need to have a new rule?

zhichao-li replied:
What's your preference? It's not a must to be a new Rule. How about embedded it into `ResolvedReferences`? I was thinking of putting this similar logic into one place.

yhuai replied:
I think `ResolvedReferences` is good.

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -322,6 +323,62 @@ class Analyzer(
+      }
+    }
+
+    def indexToColumn(
```
Can we combine this method the one above?

zhichao-li replied:
it indeed similar from the code structure and name, but internally it wants to construct two different things. And also the first parameter of `UnresolvedException` require concrete type, so it cannot be a type like TreeNode[_] to unify the api. How about I rename those to be `indexToSortOrder` and `indexToAggregate` ?

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -322,6 +323,62 @@ class Analyzer(
+            indexToColumn(s, child, index, direction)
+          case s @ SortOrder(UnaryMinus(IntegerLiteral(index)), direction) =>
+            indexToColumn(s, child, -index, direction)
+          case other => other
```
Please add comments to explain the semantic.

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -322,6 +323,62 @@ class Analyzer(
+    }
+
+    def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+      case s @ Sort(orders, global, child) if child.resolved =>
```
Can we use a narrower condition on when this case will be triggered?

zhichao-li replied:
Have thought of this for a while, but still cannot come up with a concrete solution for such nested type. Not sure if you can provide any snippet as a suggestion?

yhuai replied:
I guess that we only need to trigger this rule if there is any integer literal, right?

zhichao-li replied:
That's exactly what I trying to achieve. The thing is `expression` is hided within Seq[SortOrder], somehow we need to iterate `order` to do the checking and the conversion if match. Cann't see the chance to compact them into one line condition, and there's no performance difference in this way. 
``` scala
case class Sort(
    order: Seq[SortOrder],
    global: Boolean,
    child: LogicalPlan)
case class SortOrder(child: Expression, direction: SortDirection)
```

yhuai commented on sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:
```diff
@@ -456,21 +455,32 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
       Seq(Row(1, 2), Row(2, 2), Row(3, 2)))
 
     checkAnswer(
       sql("SELECT a, 1, sum(b) FROM testData2 GROUP BY a, 1"),
```
what is the meaning of 1 in this case?

zhichao-li replied:
`1` would be refer to the first index of columns for group by which is `a` for this case. kind of redundant here.

yhuai replied:
If we have a table with a single column, what will happen if we call `GROUP BY 1, 2`?

zhichao-li replied:
it would throw exception like "Invalid call to Aggregate by position: 2 does not exist", let me add that as an unit-test.

yhuai commented:
@zhichao-li can you also put the explanation of the semantic in the description?

zhichao-li commented:
Sure, would provide more description and comments later.

zhichao-li committed:
address comments

SparkQA commented:
**[Test build #51419 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51419/consoleFull)** for PR 10731 at commit [`66c54b1`](https://github.com/apache/spark/commit/66c54b14f8e57d181bf78f293fe0de0d899cb2ba).

SparkQA commented:
**[Test build #51419 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51419/consoleFull)** for PR 10731 at commit [`66c54b1`](https://github.com/apache/spark/commit/66c54b14f8e57d181bf78f293fe0de0d899cb2ba).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51419/
Test PASSed.

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala:
```diff
@@ -202,3 +203,14 @@ object Unions {
+object IntegerIndex {
+  def unapply(a: Any): Option[Int] = a match {
+    case Literal(a: Int, IntegerType) => Some(a)
+    case UnaryMinus(IntegerLiteral(v)) => Some(-v)
```
is it standard to support -(-1)? I see postgres support it, but somewhat strange to me.

gatorsmile replied:
This line is used for catching the illegal case:
```scala
      sql("SELECT * FROM testData2 ORDER BY -1 DESC, b ASC").collect()
```

I plan to keep it untouched in the PR. Thanks!

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -497,6 +498,41 @@ class Analyzer(
+        groups.exists(IntegerIndex.unapply(_).nonEmpty) =>
+        val newGroups = groups map {
+          case IntegerIndex(index) =>
+            val attributes = child.output
```
I'm not sure if this is correct. 1 and 2 in the group by here refers to the position 1 and 2 in the select list, not the underlying query output.

rxin replied:
we also need to check whether the position is an aggregate function. in postgres

```
rxin=# select 'one', 'two', count(*) from r1 group by 1, 3;
ERROR:  aggregate functions are not allowed in GROUP BY
LINE 1: select 'one', 'two', count(*) from r1 group by 1, 3;
                             ^
```

rxin commented:
Also I'd say "by position", not "by index", since index usually refers to something else in databases.

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -497,6 +498,41 @@ class Analyzer(
 
+      // Replace the index with the related attribute for ORDER BY
+      // which is a 1-base position of the projection list.
+      case s @ Sort(orders, global, child) if child.resolved &&
```
this rule is getting pretty long -- i wonder if there are ways to break it down

gatorsmile replied:
I will move it to the rule `ResolveSortReferences`

gatorsmile replied:
I am unable to find a good place for `group by ordinal resolution`, after placing `order by ordinal resolution` in `ResolveSortReferences`. Two options are in my mind:

- Assuming we can merge https://github.com/apache/spark/pull/11208, which splits `ResolveReferences` to two rules: `ResolveStar` and `ResolveReferences`. Then, `ResolveReferences` is not very long, maybe we can keep resolution of ordinal here.
- Create a separate rule `ResolveOrdinal` for both cases.

In the next PR, I will first pick the second option, if nobody is against it. : ) Thanks!

rxin commented:
two other comments:

1. it'd be better to separate order by position and group by position into two prs.

2. we should have config options to allow turning this off.

rxin commented:
@gatorsmile do you think you can take over this and create two prs based on this?

gatorsmile commented:
Sure, will do it. Thanks!

gatorsmile commented:
Just did a quick search. 

> SQL92 allowed the use of ordinal positions for sort_expressions, but this functionality has been deprecated and should not be used in SQL2003 queries. 

However, the mainstream RDBMS still support it.
- Oracle: https://docs.oracle.com/cd/B28359_01/server.111/b28286/statements_10002.htm#i2171079
- DB2: https://www.ibm.com/support/knowledgecenter/SSEPEK_10.0.0/com.ibm.db2z10.doc.sqlref/src/tpc/db2z_sql_orderbyclause.dita
- SQL Server: https://msdn.microsoft.com/en-us/library/ms188385.aspx

None of these top 3 enterprise RDBMS are allowing negative positions. I think we should not support the negative integer in Order by.

Thanks!

yhuai commented:
@gatorsmile Thank you for the investigation. Yea let's not use negative integer. 

Regarding group by clause, do other systems support using integer numbers?

gatorsmile commented:
@yhuai "group by position" is not supported by Oracle, DB2 and SQL Server. I am unable to find it in any SQL standard.

Should we continue to support it? Also CC @rxin

yhuai commented:
One question, if I have
```
table1: a: int, b: int, c: int

SELECT b, c FROM table1 ORDER BY 1, 2
```
what are columns used in ORDER BY? I guess `b, c`, right?

For GROUP BY, I feel it is not always obvious what are columns referred by the positions (I mean not as obvious as ORDER BY). What do you think?

gatorsmile commented:
What you said about order by is right. The most tricky part is `*`. When we are doing select (*) in DB2, the position number is based on the table definition in catalog table. 

Regarding Group By, I do not know which behavior is right. Different from Order By, Group By is below Project/SELECT. Thus, personally, I do not know what is the expected behavior when resolving position number in group by. Just like, the alias defined in Project/Select cannot be used in Group By.

yhuai commented:
Thanks. Then, let's add the support to ORDER BY.

rxin commented:
It is pretty obvious isn't it even for group by? It is just the project list, not the underlying table.

hvanhovell commented:
GROUP BY position is supported by a few major analytical databases: Terradata & Netezza

I am not sure if you should even allow the combination of a `SELECT *` with a positional ORDER BY/GROUP BY clause

rxin commented:
select * with group by is definitely not valid (with or without position)

select * with order by should work, since * here is just an expansion.

gatorsmile commented:
Just confirmed what @hvanhovell said, Netezza and Terradata support "group by position". 

- https://www.ibm.com/support/knowledgecenter/api/content/nl/en-us/SSULQD_7.2.0/com.ibm.nz.dbu.doc/r_dbuser_select.html
- http://tunweb.teradata.ws/tunstudent/TeradataUserManuals/Teradata_SQL_Quick_Reference.pdf

Also confirmed what @rxin said, in Group By, the position is based on the output columns (select expression). 

Thus, I think the integer in `groupingExpressions` should be resolved based on `aggregateExpressions` of `Aggregate`. Please let me know if my understanding is wrong. Thanks!

gatorsmile commented:
"Group By Ordinal" will throw an exception if the corresponding position of the select list is an `AggregateFunction`. This is not allowed. I believe this PR misses this point. Please correct me if my understanding is wrong. Thanks!
